### ISSUE TITLE
[question] Difference of pytorch_quantization modules and torch.ao.quantization

### ISSUE THREAD

**Issue by vadimkantorov** [NVIDIA/TensorRT#3095]
What is relation of pytorch_quantization (if I understand it's a "fake quantization" toolkit for QAT and export to ONNX for TensorRT consumption) and of PyTorch core fake quantization modules in https://pytorch.org/docs/stable/quantization-support.html ?

Is it that pytorch_quantization focuses on "fake quantization" modules and that https://pytorch.org/docs/stable/quantization-support.html#module-torch.ao.nn.qat has only a few fake quantized modules out of the box? (If I understand well, it's https://pytorch.org/docs/stable/generated/torch.ao.quantization.quantize_fx.prepare_qat_fx.html that inserts the fake quantization modules)

Thank you!

**Comment by lix19937** [comment]
I think, NV PyTorch-Quantization toolkit just for training and evaluating PyTorch models with simulated quantization. Q-DQ node(layer) can be added to the model automatically/ manually, allowing the model to be fine-tuned for accuracy and performance. Quantization is compatible with **NVIDIA int8 kernels** and TRT.    

There are some differences in the details of the quantification methods used between pytorch-quantize-fx with nv-PyTorch-Quantization.

**Comment by zerollzeng** [comment]
@ttyio for viz.

**Comment by vadimkantorov** [comment]
Because quantization is characterized by many different regimes (hard to grasp as a quantization novice):
- eager vs graph
- static vs dynamic
- ptq vs qat
- fake vs real

I would say it's important to clearly explain which modes is nv-pytorch-quantization for and how it's currently different from torch.ao.quantization

**Comment by ttyio** [comment]
Hi @vadimkantorov , we recommend nv-pytorch-quantization if you are going to deploy using TensorRT.
Here is more links for reference , the whitepaper: https://arxiv.org/pdf/2004.09602
Also we have a GTC session in https://www.nvidia.com/en-us/on-demand/session/gtcspring21-s31653/

**Comment by vadimkantorov** [comment]
Thanks, so as I understood, it's for QAT training with Fake Quantization in kind of graph/module mode (with operations done in float), targeted for further conversion and deployment with TensorRT (with int8 ops). Correct? So maybe some simple paragraph along these lines in README would clarify situation for novices :) Thanks!

**Comment by ttyio** [comment]
@vadimkantorov thank you for the suggestion. Yes we use fake quant, not only support QAT fine-tuning, we also support PTQ calibration.  There is a link to https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/ in the README, and we have document it in the link. we also mention some further optimization topic for TensorRT inference in this doc. Hope this is helpful, thanks again!

**Comment by vadimkantorov** [comment]
thank you! :)

**Comment by vadimkantorov** [comment]
@ttyio a similar question on https://discuss.pytorch.org/t/what-is-the-difference-between-pytorch-quantization-and-torch-ao-quantization/180292/5

**Comment by vadimkantorov** [comment]
@ttyio 

https://github.com/NVIDIA/TensorRT/blob/46253b644142a1d9632ba463422abfc5dcefc371/tools/pytorch-quantization/pytorch_quantization/nn/modules/tensor_quantizer.py#L312:

```python
if inputs.dtype == torch.half or amax.dtype == torch.half:
    raise Exception("Exporting to ONNX in fp16 is not supported. Please export in fp32, i.e. disable AMP.")
outputs = self._fb_fake_quant(inputs, amax)
```

It seems that this code has not changed since 2021. Is it still correct? I would think that PyTorch FakeQuant should now be supporting fp16, no?

cc @jerryzh168

**Comment by ttyio** [comment]
@vadimkantorov we will remove this in the next release which including the bfloat support. thanks!

**Comment by vadimkantorov** [comment]
We simply removed this `if`, export passed fine, but then trtexec complains with assert failure that scale is 0 or non-positive. Any hints?

**Comment by vadimkantorov** [comment]
Overall, it would be good to have an explainer about the role of pytorch_quantization in the modern PyTorch story. pytorch_quantization was conceived 3 years ago, PyTorch's built-in quantization API has improved since then...

**Comment by vadimkantorov** [comment]
@ttyio Maybe if the scale is computed in fp16, some underflow?

Then it should just replace to `amax_sequeeze.to(torch.float32) / bound.to(torch.float32)`

Also, there is a typo, `amax_sequeeze` should likely be `amax_squeeze`

**Comment by ttyio** [comment]
@vadimkantorov , yes, for the underflow/overflow issue, yes need more changes, the `amax` should cast to float buffer if low precision detected here: https://github.com/NVIDIA/TensorRT/blob/release/8.6/tools/pytorch-quantization/pytorch_quantization/nn/modules/tensor_quantizer.py#L277


for the role of nv pytorch qat tool vs. native pytorch qat, if you check the output onnx, the q/dq placement is different, and trt explicit precision network performance heavily depends on q/dq placement. https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#qdq-placement-recs

**Comment by ttyio** [comment]
@vadimkantorov btw, I see your code point to qat tool `2.1.2`, if you check the latest `release/8.6` branch, you will see latest qat `2.1.3` version.

**Comment by vadimkantorov** [comment]
@ttyio No, the same code in release/8.6: https://github.com/NVIDIA/TensorRT/blob/release/8.6/tools/pytorch-quantization/pytorch_quantization/nn/modules/tensor_quantizer.py#L279-L314 :(

It would be great to including screenshots with such Q-DQ examples and differences into docs!

**Comment by vadimkantorov** [comment]
> @ttyio Maybe if the scale is computed in fp16, some underflow?
> 
> Then it should just replace to `amax_sequeeze.to(torch.float32) / bound.to(torch.float32)`
> 
> Also, there is a typo, `amax_sequeeze` should likely be `amax_squeeze`

@ttyio btw this seems to help: `amax_sequeeze.to(torch.float32) / float(bound)`

**Comment by ttyio** [comment]
@vadimkantorov the fp16 onnx export is not supported in 2.1.3, we used gpu kernel to accelerate the fine-tuning so I point out the branch. 

Thank you for your suggesting we will continue improving the document.

For the float16 problem, TRT is weak typed, it does not matter what the output of gemm is fp32 or fp16, we use builder flag to allow fp16 kernel https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/BuilderConfig.html#tensorrt.BuilderFlag

**Comment by vadimkantorov** [comment]
In general, what was the basis for those constraints against fp16? If we still want to do it, do we just need to patch the onnx file?

(Or is the next release supporting fp16/opset19 just around the corner?)

**Comment by ttyio** [comment]
>> but there still was inserted a reformat node to convert back from fp32 output.
 @vadimkantorov , this is not caused by fp16/opset19, TRT is weak typed till version 8.6. Is this layer mark as network output? or the next layer run in FP32?

**Comment by ttyio** [comment]
@vadimkantorov  network out is FP32 by default, to remove the reformat, change the network output to Fp16 follow this sample: https://github.com/NVIDIA/TensorRT/tree/release/8.6/samples/sampleIOFormats

**Comment by 1438802682** [comment]
Hello, I apologize for disturbing you in your free time. May I ask what is the highest version of PyTorch that pytorch_quantization supports?

**Comment by lix19937** [comment]
@1438802682 Maybe support v1.13, you can try it.

**Comment by ttyio** [comment]
> Hello, I apologize for disturbing you in your free time. May I ask what is the highest version of PyTorch that pytorch_quantization supports?

we tested in

**Comment by Michelvl92** [comment]
> Hi @vadimkantorov , we recommend nv-pytorch-quantization if you are going to deploy using TensorRT. Here is more links for reference , the whitepaper: https://arxiv.org/pdf/2004.09602 Also we have a GTC session in https://www.nvidia.com/en-us/on-demand/session/gtcspring21-s31653/

Does this still hold in February 2024 when deploying to TensorRT? nv-pytorch-quantization 2.2.0 gives the warning to move to PyTorch naive quantization, but, there is no documentation on [PyTorch quantization documentation](https://pytorch.org/docs/stable/quantization.html) focusing on deployment to ONNX or TensorRT, but focus only on CPU deployment.
